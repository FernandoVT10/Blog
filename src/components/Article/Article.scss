@use "../../styles/variables";

.article {
    background: variables.$primary;
    padding: 0;
    color: white;

    &__cover {
        width: 100%;
        height: 400px;
        background-position: center !important;
        background-attachment: fixed !important;
        background-size: cover !important;

        &--edit {
            position: relative;
            display: flex;
            justify-content: center;
            align-items: center;

            &::before {
                content: "";
                opacity: 0;
                position: absolute;
                width: 100%;
                height: 100%;
                background: rgba($color: #000000, $alpha: .8);
                transition: .3s;
            }
        }

        &__label {
            cursor: pointer;
            position: relative;
            opacity: 0;
            z-index: 2;
            color: white;
            font-size: 30px;
            padding: 15px 40px;
            background: variables.$primary;
            transition: .3s;
        }

        &__input {
            display: none;
        }

        &:hover {
            &::before {
                opacity: 1;
            }

            & .article__cover__label {
                opacity: 1;

                &:hover {
                    opacity: .6;
                }
            }
        }
    }

    &__title {
        font-weight: 600;
        font-size: 30px;
    }

    &__edit-title {
        outline: 0;
        width: 100%;
        font-weight: 600;
        font-size: 30px;
        border: none;
        color: white;
        height: 50px;
        resize: none;
        word-break: break-all;
        background: transparent;

        &-error {
            display: none;
            color: variables.$danger;
            margin-bottom: 15px;
        }

        &:invalid {
            box-shadow: 0 0px 2px 1px variables.$danger;

            & ~ .article__edit-title-error {
                display: block;
            }
        }
    }

    &__content-container {
        padding: 30px;
    }

    &__content {
        margin-top: 30px;
    }

    &__content-editor {
        margin-bottom: 15px;

        &.ql-disabled {
            border: none !important;
        }
    }

    &__description {
        width: 100%;
        padding: 20px;
        border-radius: 5px;
        background: transparent;
        color: white;
    }

    &__categories {
        margin-top: 10px;
        font-weight: bold;
        font-size: 14px;
        color: variables.$secondary;

        &__container {
            display: inline;
        }

        &__button-container {
            display: inline-block;
            position: relative;
            margin: 0 5px;
        }

        &__add-category {
            z-index: 10;
            position: absolute;
            display: none;
            top: 30px;
            left: 0;
            flex-direction: column;
            background: variables.$black;
            border: solid 2px variables.$secondary;
            padding: 10px 20px;
            max-height: 160px;
            width: 180px;
            overflow: auto;

            &--active {
                display: flex;
            }

            &::-webkit-scrollbar {
                width: 3px;
                background: variables.$black;
            }

            &::-webkit-scrollbar-thumb {
                background: variables.$secondary;
            }
        }

        &__add-button {
            position: relative;
            outline: 0 !important;
            background: transparent;
            font-size: 12px;
            border: none;
            border-radius: 5px;
            color: variables.$secondary;
            transition: .3s;

            &:hover {
                color: white;
            }
        }
    }

    &__category {
        display: inline-block;
        color: variables.$secondary;
        transition: .3s;

        &:hover {
            text-decoration: none;
            color: white;
        }
    }

    &__buttons-container {
        position: fixed;
        z-index: 100;
        right: 25px;
        bottom: 25px;
    }

    &__float-button {
        outline: 0 !important;
        display: inline-flex;
        margin: 10px;
        justify-content: center;
        align-items: center;
        background: variables.$secondary;
        width: 50px;
        height: 50px;
        border-radius: 50%;
        border: none;
        transition: .3s;

        &--cancel {
            background: variables.$danger;
            color: white;
        }

        &:hover {
            opacity: .8;
        }
    }
}

// Quill editor styles

.ql-snow {
    & .ql-stroke {
        stroke: white;
        transition: .3s;
    }

    & .ql-fill {
        fill: white;
        transition: .3s;
    }

    & .ql-picker-label {
        color: white;

        &:hover {
            color: variables.$secondary;
        }
    }

    & .ql-picker-options {
        border: none !important;
        padding: 20px;
        background: variables.$black;
    }

    & .ql-picker-item {
        color: white;
        transition: .3s;

        &.ql-selected {
            color: variables.$secondary;
        }
    }

    & p {
        line-height: 40px;
        font-size: 18px;
        margin: 0;
    }

    &.ql-toolbar {
        &.disabled {
            display: none;
        }

        & .ql-picker-label:hover,
        & .ql-picker-item.ql-selected,
        & .ql-picker-item:hover {
            color: variables.$secondary;
        }

        & button:hover,
        & button:focus,
        & button.ql-active {
            & .ql-stroke {
                stroke: variables.$secondary;
            }
        
            & .ql-fill {
                fill: variables.$secondary;
            }
        }
    }

    & .ql-syntax {
        font-weight: bold;
        font-family: consolas;
        font-size: 18px;
    }
}