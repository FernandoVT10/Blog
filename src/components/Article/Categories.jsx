import Api from "../../ApiController";

import Link from "next/link";
import { useState, useEffect } from "react";

export default ({ editArticle, categories, onChangeCategories }) => {
    const [selectedCategories, setSelectedCategories] = useState([]);
    const [allCategories, setAllCategories] = useState([]);
    const [addCategoryStatus, setAddCategoryStatus] = useState(false);

    useEffect(() => {
        Api.get("categories/getAllCategories")
        .then(data => {
            if(data.length) {
                setAllCategories(data.map(category => {
                    let active = false;

                    categories.forEach(({ name }) => {
                        if(name === category.name) {
                            active = true;
                        }
                    });

                    return {
                        _id: category._id,
                        name: category.name,
                        active
                    };
                }));
            }
        });
    }, []);

    useEffect(() => setSelectedCategories(categories.map(
        category => category.name
    )), [categories]);

    const handleCheckbox = (categoryName) => {
        if(selectedCategories.includes(categoryName)) {
            const newCategories = selectedCategories.filter(
                localCategory => localCategory !== categoryName
            );

            setSelectedCategories(newCategories);
            onChangeCategories(newCategories);
        } else {
            const newCategories = [...selectedCategories, categoryName];

            onChangeCategories(newCategories);
            setSelectedCategories(newCategories);
        }

        setAllCategories(allCategories.map(allCategory => {
            if(allCategory.name === categoryName) {
                return (
                    {
                        _id: allCategory._id,
                        name: allCategory.name,
                        active: !allCategory.active
                    }
                );
            }
            
            return allCategory;
        }));
    };

    const getCategories = () => {
        return allCategories.map(category => {
            return (
                <div key={category._id} className="formulary__checkbox">
                    <input
                    className="formulary__checkbox__input"
                    id={`category-${category.name}`}
                    onChange={() => handleCheckbox(category.name)}
                    type="checkbox"
                    checked={category.active} />

                    <label
                    className="formulary__checkbox__label"
                    htmlFor={`category-${category.name}`}>
                        <span className="formulary__checkbox__label-checkbox"></span>
                        { category.name }
                    </label>
                </div>
            );
        });
    };

    if(editArticle) {
        const addCategoryClass = addCategoryStatus
        ? "article__categories__add-category--active"
        : "";

        return (
            <div className="article__categories">
                <i className="fas fa-tags" aria-hidden="true"></i>

                <div
                className="article__categories__button-container">
                    <button
                    className="article__categories__add-button"
                    onClick={() => setAddCategoryStatus(!addCategoryStatus)}>
                        <i className="fas fa-plus" aria-hidden="true"></i>
                    </button>

                    <div className={`article__categories__add-category ${addCategoryClass}`}>
                        { getCategories() }
                    </div>
                </div>

                <div className="article__categories__container">
                    {selectedCategories.map((categoryName, index) => {
                        return (
                            <Link
                            href={`/articles?categories=${categoryName}`}
                            key={index}>
                                <a className="article__category mr-2">
                                    { categoryName }
                                </a>
                            </Link>
                        );
                    })}
                </div>
            </div>
        );
    }

    if(categories.length) {
        return (
            <div className="article__categories">
                <i className="fas fa-tags" aria-hidden="true"></i>

                <span className="ml-2">
                    {categories.map((category, index) => {
                        return (
                            <Link
                            href={`/articles?categories=${category.name}`}
                            key={index}>
                                <a className="article__category mr-2">
                                    { category.name }
                                </a>
                            </Link>
                        );
                    })}
                </span>
            </div>
        );
    }

    return null;
};