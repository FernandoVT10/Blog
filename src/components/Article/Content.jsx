import { useEffect, useRef } from "react";
import Api from "../../ApiController";

export default ({ editArticle, content, onChangeContent }) => {
    const editor = useRef();

    useEffect(() => {
        async function setupQuill() {
            hljs.configure({
                languages: ["javascript"]
            });

            const quill = await import("quill");

            editor.current = new quill.default("#content-editor", {
                modules: {
                    syntax: true,
                    toolbar: [
                        [{ header: [1, 2, false] }],
                        ["bold", "italic", "underline", "link"],
                        ["image", "code-block"],
                        [{ list: "ordered"}, { list: "bullet" }]
                    ]
                },
                readOnly: !editArticle,
                imageHandler,
                theme: "snow"
            });

            function imageHandler() {
                const input = document.createElement("input");
                input.setAttribute("type", "file");
                input.click();

                input.onchange = () => {
                    const file = input.files[0];

                    if(file.type === "image/png"
                    || file.type === "image/jpg"
                    || file.type === "image/jpeg"
                    || file.type === "image/gif") {
                        const formData = new FormData();
                        formData.append("image", file);

                        Api.post("images/uploadImage", formData, true, true)
                        .then(data => {
                            if(data.status) {
                                const range = editor.current.getSelection();
                                
                                editor.current.insertEmbed(
                                    range.index,
                                    "image",
                                    `http://localhost:3000/img/articles/content/${data.imageName}`
                                );
                            }
                        });
                    }
                };
            }

            editor.current.getModule("toolbar").addHandler("image", imageHandler);

            editor.current.on("text-change", () => {
                const content = JSON.stringify(editor.current.getContents().ops);

                onChangeContent(content);
            });

            if(!editArticle) {
                const toolbar = document.querySelector(".ql-toolbar");
                toolbar.classList.add("disabled");
            }

            try {
                editor.current.setContents(JSON.parse(content));
            } catch (error) { }
        }

        setupQuill();
    }, []);

    useEffect(() => {
        if(editor.current) {
            const toolbar = document.querySelector(".ql-toolbar");
            
            if(editArticle) {
                // if the edit article is true, we enable the editor
                editor.current.enable();
                // and we enable the editor toolbar
                toolbar.classList.remove("disabled");
            } else {
                // if the edit article is false, we disable the editor
                editor.current.disable();
                // we hide the editor toolbar
                toolbar.classList.add("disabled");
            }

            try {
                editor.current.setContents(JSON.parse(content));
            } catch (error) { }
        }
    }, [editArticle]);

    return (
        <div>
            <div id="content-editor" className="article__content-editor">
            </div>
        </div>
    );
};