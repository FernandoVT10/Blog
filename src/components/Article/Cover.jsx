import { useState, useCallback, useEffect } from "react";

export default ({ editArticle, onChangeImage, cover }) => {
    const [imageURL, setImageURL] = useState("");

    useEffect(() => {
        setImageURL(`/img/articles/${cover}`);
    }, [editArticle]);

    const changeImage = useCallback(({ target: { files } }) => {
        const file = files[0];

        if(file.type === "image/png"
        || file.type === "image/jpeg"
        || file.type === "image/jpg"
        || file.type === "image/gif") {
            onChangeImage(file);
            setImageURL(URL.createObjectURL(file));
        }
    });

    if(editArticle) {
        return (
            <div
            className="article__cover article__cover--edit"
            style={{
                background: `url(${imageURL})`
            }}>
                <label
                htmlFor="cover-image"
                className="article__cover__label">
                    <i className="fas fa-image" aria-hidden="true"></i>
                </label>

                <input
                type="file"
                id="cover-image"
                onChange={changeImage}
                className="article__cover__input" />
            </div>
        );
    }

    return (
        <div
        className="article__cover"
        style={{
            background: `url(${imageURL})`
        }}></div>
    );
};