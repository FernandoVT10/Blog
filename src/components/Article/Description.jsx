import React from "react";

export default ({ editArticle, onChangeDescription, description }) => {
    const onChange = ({ target: { value } }) => onChangeDescription(value);

    if(!editArticle) {
        return null;
    }

    return (
        <textarea
        className="article__description"
        placeholder="Enter a description"
        maxLength="250"
        onChange={onChange}
        rows="3">{ description }</textarea>
    );
}