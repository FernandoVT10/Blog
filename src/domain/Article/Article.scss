.article {
    background: var(--primary);
    padding: 0;
    margin-top: 30px;
    color: var(--white);
    
    &__cover {
        width: 100%;
        height: 400px;
        background-position: center !important;
        background-attachment: fixed !important;
        background-size: cover !important;
    }

    &__content-container {
        padding: 30px;
    }

    &__title {
        font-weight: 600;
        font-size: 2em;
    }

    &__categories {
        margin-top: 10px;
        font-weight: bold;
        font-size: 0.8em;
        color: var(--secondary);
    }

    &__category {
        display: inline-block;
        color: var(--secondary);
        transition: .3s;

        &:hover {
            text-decoration: none;
            color: var(--white);
        }
    }

    &__loader {
        display: flex;
        align-items: center;
        justify-content: center;
        background: var(--primary);
        color: var(--secondary);
        margin-top: 30px;
    }

    &__not-found {
        display: flex;
        flex-direction: column;
        justify-content: center;
        align-items: center;
        color: var(--secondary);
        background: var(--primary);
        font-weight: bold;
        margin-top: 30px;
        font-size: 1.5em;
    }

    &__404 {
        font-size: 5em;
        color: var(--white);
    }
}