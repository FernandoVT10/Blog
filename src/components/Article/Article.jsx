import Cover from "./Cover";
import Title from "./Title";
import Content from "./Content";
import Categories from "./Categories";
import Description from "./Description";
import Api from "../../ApiController";

import { useState, useEffect } from "react";

import "./Article.scss";

export default (props) => {
    const [coverImage, setCoverImage] = useState();
    const [title, setTitle] = useState("");
    const [content, setContent] = useState("");
    const [description, setDescription] = useState("");
    const [categories, setCategories] = useState([]);

    const [article, setArticle] = useState(props.article);
    const [editArticle, setEditArticle] = useState(false);

    useEffect(() => {
        setTitle(article.title);
        setContent(article.content);
        setDescription(article.description);
        setCategories(article.categories.map(
            category => category.name
        ));
    }, []);

    const saveArticle = () => {
        if(!title.length) {
            return;
        }

        const formData = new FormData();

        formData.append("articleId", article._id);
        formData.append("cover", coverImage);
        formData.append("title", title);
        formData.append("content", content);
        formData.append("description", description);
        formData.append("categories", categories);

        Api.post("articles/updateArticle/", formData, true, true)
        .then(res => {
            if(res.status) {
                const newArticle = res.newArticle;

                setArticle(newArticle);
            }
            setEditArticle(false);
        });
    };

    const getFloatButtons = () => {
        if(editArticle) {
            return (
                <div className="article__buttons-container">
                    <button
                    className="article__float-button"
                    onClick={saveArticle}>
                        <i className="fas fa-folder-open" aria-hidden="true"></i>
                    </button>

                    <button
                    className="article__float-button article__float-button--cancel"
                    onClick={() => setEditArticle(false)}>
                        <i className="fas fa-times" aria-hidden="true"></i>
                    </button>
                </div>
            );
        }

        return (
            <div className="article__buttons-container">
                <button
                className="article__float-button"
                onClick={() => setEditArticle(true)}>
                    <i className="fas fa-pencil-alt" aria-hidden="true"></i>
                </button>
            </div>
        );
    };

    return (
        <div className="container-fluid">
            <div className="row">
                <div className="col-12 article">
                    <Cover
                    editArticle={editArticle}
                    cover={article.cover}
                    onChangeImage={setCoverImage}/>

                    <div className="article__content-container">
                        <Title
                        editArticle={editArticle}
                        title={article.title}
                        onChangeTitle={setTitle}/>
                        
                        <Content
                        editArticle={editArticle}
                        content={article.content}
                        onChangeContent={setContent}/>

                        <Description
                        editArticle={editArticle}
                        description={description}
                        onChangeDescription={setDescription}/>

                        <Categories
                        editArticle={editArticle}
                        categories={article.categories}
                        onChangeCategories={setCategories}/>
                    </div>
                </div>

                { getFloatButtons() }
            </div>
        </div>
    );
};