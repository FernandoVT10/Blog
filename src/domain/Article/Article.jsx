import Content from "./Content";

import ApiController from "../../services/ApiController";

import { useState, useEffect } from "react";
import { useRouter } from "next/router";
import Link from "next/link";

import "./Article.scss";

export default () => {
    const [article, setArticle] = useState({});
    const [loading, setLoading] = useState(true);

    const router = useRouter();

    useEffect(() => {
        const { articleId } = router.query;

        if(articleId && loading) {
            ApiController.get(`articles/${articleId}`)
            .then(res => {
                if(res.data) {
                    const { title } = res.data.article;
                    window.document.title = `${title} - Fernando Vaca Tamayo Blog`;

                    setArticle(res.data.article);
                }

                setLoading(false);
            });
        }
    }, [router.query]);

    const getCategories = () => {
        if(article.categories.length) {
            return (
                <div className="article__categories">
                    <i className="fas fa-tags" aria-hidden="true"></i>

                    <span className="ml-2">
                        {article.categories.map((category, index) => {
                            return (
                                <Link
                                href={`/articles?categories=${category.name}`}
                                key={index}>
                                    <a className="article__category mr-2">
                                        { category.name }
                                    </a>
                                </Link>
                            );
                        })}
                    </span>
                </div>
            );
        }
    }

    if(!article.title && !loading) {
        return (
            <div className="container-fluid">
                <div className="row">
                    <div className="col-12 body article__not-found">
                        <span className="article__404">404</span>
                        The article doesn't exist.
                    </div>
                </div>
            </div>
        );
    }

    if(loading) {
        return (
            <div className="container-fluid">
                <div className="row">
                    <div className="col-12 article__loader body">
                        <span className="spinner-border"></span>
                    </div>
                </div>
            </div>
        );
    }

    return (
        <div className="container-fluid">
            <div className="row">
                <div className="col-12 article">
                    <div
                    className="article__cover"
                    style={{ background: `url(/img/articles/${article.cover})` }}></div>

                    <div className="article__content-container">
                        <h1 className="article__title">
                            { article.title }
                        </h1>

                        <Content content={article.content}/>

                        { getCategories() }
                    </div>
                </div>
            </div>
        </div>
    );
};