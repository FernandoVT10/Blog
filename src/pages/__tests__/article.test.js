import { act, Simulate } from "react-dom/test-utils";
import { render } from "react-dom";
import Article from "../article/[articleId]";

const ARTICLE_MOCK = {
    _id: 18,
    title: "Test Article",
    description: "Test description",
    cover: "cover.jpg",
    content: "Test content",
    categories: [
        { name: "React JS" },
        { name: "Node JS" }
    ]
};

const COMMENT_MOCKS = [
    {
        _id: 1,
        name: "Test name",
        comment: "This is a test"
    },
    {
        _id: 2,
        name: "Test name 2",
        comment: "This is a test"
    },
    {
        _id: 3,
        name: "Test name 3",
        comment: "This is a test"
    }
];

let container;

describe("Index page", () => {
    beforeEach(() => {
        fetchMock.doMock();

        fetchMock.mockOnce(JSON.stringify(
            COMMENT_MOCKS
        ));

        container = document.createElement('div');
        document.body.appendChild(container);
    });
    
    afterEach(() => {
        document.body.removeChild(container);
        container = null;
    });

    it('Check if renders correctly', async () => {
        await act(async () => {
            render(<Article article={ARTICLE_MOCK}/>, container);
        });

        expect(container).toMatchSnapshot();
    });

    it('It should load more comments', async () => {
        await act(async () => {
            render(<Article article={ARTICLE_MOCK}/>, container);
        });

        const button = container.querySelector(".comments__load-more");

        fetchMock.once(JSON.stringify(
            {
                _id: 4,
                name: "Test name 4",
                comment: "This is a test 4"
            }
        ));

        await act(async () => {
            Simulate.click(button); 
        });

        const comments = container.querySelectorAll(".comment");

        const commentName = comments[3].querySelector(".comment__name").textContent;
        const commentContent = comments[3].querySelector(".comment__content").textContent;

        expect(comments.length).toBe(4);
        expect(commentName).toBe("Test name 4");
        expect(commentContent).toBe("This is a test 4");
    });

    it('It should create a comment', async () => {
        await act(async () => {
            render(<Article article={ARTICLE_MOCK}/>, container);
        });

        const input = container.querySelector("input");
        const textarea = container.querySelector("textarea");
        const form = container.querySelector("form");

        fetchMock.once(JSON.stringify(
            {
                status: true,
                createdComment: {
                    _id: 45,
                    name: "Test name 45",
                    comment: "This is a test 45"
                }
            }
        ));

        act(() => {
            input.value = "Test name 45";
            textarea.value = "This is a test 45";

            Simulate.change(input);
            Simulate.change(textarea);
        });

        await act(async () => {
            Simulate.submit(form); 
        });

        const comments = container.querySelectorAll(".comment");

        const commentName = comments[0].querySelector(".comment__name").textContent;
        const commentContent = comments[0].querySelector(".comment__content").textContent;

        expect(comments.length).toBe(4);
        expect(commentName).toBe("Test name 45");
        expect(commentContent).toBe("This is a test 45");
    });
});