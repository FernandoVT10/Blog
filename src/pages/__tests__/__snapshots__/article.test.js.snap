// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index page Check if renders correctly 1`] = `
<div>
  <div>
    <header>
      <nav
        class="navbar navbar-expand-md navbar-dark"
      >
        <a
          class="navbar-brand navbar__brand"
          href="/"
        >
          <img
            alt="Icon"
            class="d-inline-block align-top rounded-circle"
            height="50"
            src="/favicon.ico"
            width="50"
          />
          <h1
            class="navbar__title"
          >
            Fernando Vaca Tamayo
          </h1>
        </a>
        <button
          aria-controls="main-navbar"
          aria-expanded="false"
          aria-label="Toggle navigation"
          class="navbar-toggler"
          data-target="#main-navbar"
          data-toggle="collapse"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          class="collapse navbar-collapse navbar__items"
          id="main-navbar"
        >
          <ul
            class="navbar-nav"
          >
            <li
              class="nav-item"
            >
              <a
                class="nav-link navbar__link"
                href="/"
              >
                Home
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link navbar__link"
                href="/articles"
              >
                Articles
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link navbar__link"
                href="/projects"
              >
                Projects
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link navbar__link"
                href="/contact-me"
              >
                Contact me
              </a>
            </li>
          </ul>
        </div>
      </nav>
    </header>
    <div
      class="container-fluid"
    >
      <div
        class="row"
      >
        <div
          class="col-12 article"
        >
          <img
            alt="Test Article"
            class="article__cover"
            src="/img/articles/cover.jpg"
          />
          <div
            class="article__content-container"
          >
            <h1
              class="article__title"
            >
              Test Article
            </h1>
            <div
              class="article__content"
            >
              Test content
            </div>
            <div
              class="article__categories"
            >
              <i
                aria-hidden="true"
                class="fas fa-tags"
              />
              <span
                class="ml-2"
              >
                <span>
                  React JS
                </span>
                <span>
                  , 
                  Node JS
                </span>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="container-fluid"
    >
      <div
        class="full-screen-loader "
      >
        <div
          class="full-screen-loader__spinner spinner-border text-light"
          role="status"
        >
          <span
            class="sr-only"
          >
            Loading...
          </span>
        </div>
        <h3
          class="full-screen-loader__text"
        >
          Loading...
        </h3>
      </div>
      <div
        class="row comments"
      >
        <div
          class="col-12"
        >
          <h3
            class="comments__title"
          >
            Comments
          </h3>
        </div>
        <div
          class="col-12 col-lg-8"
        >
          <form>
            <input
              class="formulary__input mb-2"
              maxlength="30"
              placeholder="Enter your Name"
              required=""
              type="text"
            />
            <textarea
              class="formulary__textarea"
              maxlength="500"
              placeholder="Enter your Message"
              required=""
              rows="6"
            />
            <button
              class="submit-button mt-3"
              type="submit"
            >
              Add Comment
            </button>
          </form>
        </div>
        <div
          class="col-12 col-lg-8"
        >
          <div
            class="comment"
          >
            <p
              class="comment__name"
            >
              Test name
            </p>
            <p
              class="comment__content"
            >
              This is a test
            </p>
          </div>
          <div
            class="comment"
          >
            <p
              class="comment__name"
            >
              Test name 2
            </p>
            <p
              class="comment__content"
            >
              This is a test
            </p>
          </div>
          <div
            class="comment"
          >
            <p
              class="comment__name"
            >
              Test name 3
            </p>
            <p
              class="comment__content"
            >
              This is a test
            </p>
          </div>
        </div>
        <div
          class="col-12 col-lg-8 d-flex justify-content-center"
        >
          <button
            class="comments__load-more "
          >
            Load more comments
          </button>
        </div>
      </div>
    </div>
    <footer
      class="main-footer container-fluid mt-4"
    >
      <div
        class="full-screen-loader "
      >
        <div
          class="full-screen-loader__spinner spinner-border text-light"
          role="status"
        >
          <span
            class="sr-only"
          >
            Loading...
          </span>
        </div>
        <h3
          class="full-screen-loader__text"
        >
          Loading...
        </h3>
      </div>
      <div
        class="modal fade"
        id="-modal"
        role="dialog"
        tabindex="-1"
      >
        <div
          class="modal-dialog "
          role="document"
        >
          <div
            class="modal-content modal__content"
          >
            <div
              class="modal-header modal__header"
            >
              <h5
                class="modal__title"
              >
                Suscription confirm status
              </h5>
              <button
                aria-label="Close"
                class="close modal__button"
                data-dismiss="modal"
                type="button"
              >
                <span
                  aria-hidden="true"
                >
                  Ã—
                </span>
              </button>
            </div>
            <div
              class="modal-body modal__body"
            >
              <p />
            </div>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-12 col-sm-6 col-xl-3"
        >
          <a
            class="main-footer__brand"
            href="/"
          >
            <img
              alt="Icon"
              class="d-inline-block align-top rounded-circle"
              height="50"
              src="/favicon.ico"
              width="50"
            />
            <h3
              class="main-footer__title"
            >
              Fernando Vaca Tamayo
            </h3>
          </a>
          <div
            class="mt-4"
          >
            <h3
              class="main-footer__subtitle"
            >
              Contact Me
            </h3>
            <p>
              <i
                aria-hidden="true"
                class="fas fa-envelope main-footer__icon"
              />
              <span
                class="main-footer__text"
              >
                vacatamayo321@gmail.com
              </span>
            </p>
          </div>
        </div>
        <div
          class="col-12 col-sm-6 col-xl-3"
        >
          <h3
            class="main-footer__subtitle"
          >
            Social Networks
          </h3>
          <a
            href="https://twitter.com/FernandoVT10"
            target="_blank"
          >
            <p>
              <i
                aria-hidden="true"
                class="fab fa-twitter main-footer__icon"
              />
              <span
                class="main-footer__text"
              >
                @FernandoVT10
              </span>
            </p>
          </a>
          <a
            href="https://github.com/FernandoVT10"
            target="_blank"
          >
            <p>
              <i
                aria-hidden="true"
                class="fab fa-github main-footer__icon"
              />
              <span
                class="main-footer__text"
              >
                FernandoVT10
              </span>
            </p>
          </a>
        </div>
        <div
          class="col-12 col-sm-6 col-xl-3"
        >
          <h3
            class="main-footer__subtitle"
          >
            Description
          </h3>
          <p
            class="main-footer__description"
          >
            This blog is my personal blog xD
          </p>
        </div>
        <div
          class="col-12 col-sm-6 col-xl-3"
        >
          <h3
            class="main-footer__subtitle"
          >
            Suscribe to my blog
          </h3>
          <form>
            <div
              class="formulary__input-container"
            >
              <input
                autocomplete="email"
                class="formulary__input "
                id="-input"
                maxlength="100"
                placeholder="Enter your email"
                required=""
                type="email"
              />
              <label
                class="formulary__error-label mt-2 "
                for="-input"
              >
                <i
                  aria-hidden="true"
                  class="fas fa-info-circle mr-1"
                />
                
              </label>
            </div>
            <button
              class="submit-button mt-3"
            >
              Suscribe
            </button>
          </form>
        </div>
      </div>
    </footer>
    <script
      src="https://cdn.jsdelivr.net/npm/chart.js@2.9.3/dist/Chart.min.js"
    />
    <script
      src="https://code.jquery.com/jquery-3.4.1.slim.min.js"
    />
    <script
      src="https://cdn.jsdelivr.net/npm/popper.js@1.16.0/dist/umd/popper.min.js"
    />
    <script
      src="https://stackpath.bootstrapcdn.com/bootstrap/4.4.1/js/bootstrap.min.js"
    />
    <script
      src="https://kit.fontawesome.com/63ef8f1397.js"
    />
  </div>
</div>
`;
